contracts/modules/RelayerModule.sol

```diff
@@ -0,0 +1,372 @@
+        uint256 requiredSignatures,
+        OwnerSignature ownerSignatureRequirement
+    )
+        internal
+        returns (bool)
+    {
+        if (requiredSignatures == 1 && ownerSignatureRequirement == OwnerSignature.Required) {
+            // use the incremental nonce
+            if (_nonce <= relayer[_wallet].nonce) {
+                return false;
+            }
+            uint256 nonceBlock = (_nonce & 0xffffffffffffffffffffffffffffffff00000000000000000000000000000000) >> 128;
+            if (nonceBlock > block.number + BLOCKBOUND) {
+                return false;
+            }
+            relayer[_wallet].nonce = _nonce;
+            return true;
+        } else {
+            // use the txHash map
+            if (relayer[_wallet].executedTx[_signHash] == true) {
+                return false;
+            }
+            relayer[_wallet].executedTx[_signHash] = true;
+            return true;
+        }
+    }
+
+    /**
+    * @dev Validates the signatures provided with a relayed transaction.
+    * The method MUST throw if one or more signatures are not valid.
+    * @param _wallet The target wallet.
+    * @param _signHash The signed hash representing the relayed transaction.
+    * @param _signatures The signatures as a concatenated byte array.
+    * @param _option An enum indicating whether the owner is required, optional or disallowed.
+    * @return A boolean indicating whether the signatures are valid.
+    */
+    function validateSignatures(
+        address _wallet,
+        bytes32 _signHash,
+        bytes memory _signatures,
+        OwnerSignature _option
+    )
+        internal
+        view
+        returns (bool)
+    {
+        if (_signatures.length == 0) {
+            return true;
+        }
+        address lastSigner = address(0);
+        address[] memory guardians;
+        if (_option != OwnerSignature.Required || _signatures.length > 65) {
+            guardians = guardianStorage.getGuardians(_wallet); // guardians are only read if they may be needed
+        }
+        bool isGuardian;
+
+        for (uint8 i = 0; i < _signatures.length / 65; i++) {
+            address signer = Utils.recoverSigner(_signHash, _signatures, i);
+
+            if (i == 0) {
+                if (_option == OwnerSignature.Required) {
+                    // First signer must be owner
+                    if (isOwner(_wallet, signer)) {
+                        continue;
+                    }
+                    return false;
+                } else if (_option == OwnerSignature.Optional) {
+                    // First signer can be owner
+                    if (isOwner(_wallet, signer)) {
+                        continue;
+                    }
+                }
+            }
+            if (signer <= lastSigner) {
+                return false; // Signers must be different
+            }
+            lastSigner = signer;
+            (isGuardian, guardians) = GuardianUtils.isGuardian(guardians, signer);
```

Comment:
it returns `true and the list of guardians minus the found guardian upon success, false and the original list of guardians if not found`

Reply:
Not sure I understand your comment. `GuardianUtils.isGuardian` returns a boolean and the updated memory array of guardians.

Reply:
Sorry, I got confused by the other two functions with the same name in `GuardianStorage.isGuardian` and `GuardianManager.isGuardian`, plus the fact that I expect a function called `isGuardian` to just return bool. But also can't find an instance where the `guardians` returned is actually used. I'd like us to have just a single `isGuardian` method but since it's not the job of this PR I'll add a backlog item if you're both okay with that?

Reply:
The `guardians` array is updated by the `isGuardian` method and used in the next iteration of the loop. Agree to add an improve ticket in the backlog.